src/dbnode/x/xio/segment_reader.go

```diff
@@ -46,40 +47,101 @@ func (sr *segmentReader) Clone(
 	return NewSegmentReader(sr.segment.Clone(pool)), nil
 }
 
-func (sr *segmentReader) Read(b []byte) (int, error) {
-	if len(b) == 0 {
-		return 0, nil
+func (sr *segmentReader) Read64() (word uint64, n byte, err error) {
+	sr.lazyInit()
+
+	var (
+		nh    = len(sr.lazyHead)
+		nht   = nh + len(sr.lazyTail)
+		res   uint64
+		bytes byte
+	)
+
+	if sr.si >= nht {
+		return 0, 0, io.EOF
 	}
 
-	if b := sr.segment.Head; b != nil && len(sr.lazyHead) == 0 {
-		sr.lazyHead = b.Bytes()
+	if sr.si+8 < nh {
```

Comment:
I believe this should be:

```suggestion
	if sr.si+8 <= nh {
```

Also a few more similar cases below.

Reply:
Good catch, will save a few cycles.